### PR TITLE
stack: add SwitchReady service

### PR DESCRIPTION
**Pull request by hawkw**
This is essentially a form of fallback that falls back when a primary
service takes too long to become ready. When the primary service's
`poll_ready` does not return ready for over the specified duration, the
secondary service is used until the primary becomes ready again,
resetting the readiness timeout. Erros from both services are
propagated.

I considered naming this `ReadyFallback`, but that name was already used
for an unrelated type in the `fallback` module. :)

### CODE REVIEW

File: linkerd/stack/src/switch_ready.rs [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::time::{delay_for, Delay, Instant};
+
+#[derive(Debug, Clone)]
+pub struct SwitchReadyLayer<B> {
```
Comment (olix0r) [inline linkerd/stack/src/switch_ready.rs:11]: For now, let's skip the impl of SwitchReadyLayer -- this is something we can easily do via layer::mk as needed for now, and start with a `SwitchReady::new` instead.

**Review by olix0r — CHANGES_REQUESTED** [review]
This looks really good! Thanks for tests.

* Let's skip the layer impl for now and instead expose a SwitchReady::new
* Let's base this directly off of `main` so it can merge independently?

File: linkerd/stack/src/switch_ready.rs (hunk 1 above)
Comment (hawkw) [reply linkerd/stack/src/switch_ready.rs:11]: yup, i considered using `layer::mk`, and can happily make that change now.

File: linkerd/stack/src/switch_ready.rs [hunk 2]
```diff
@@ -0,0 +1,389 @@
+use super::NewService;
+use futures::future::Either;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::time::{delay_for, Delay, Instant};
+/// A service which falls back to a secondary service if the primary service
```
Comment (olix0r) [inline linkerd/stack/src/switch_ready.rs:10]: newline, please

File: linkerd/stack/src/switch_ready.rs [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+            delay: delay_for(switch_after),
+            state: State::Primary,
+        }
+    }
+}
+
+impl<A, B, R> tower::Service<R> for SwitchReady<A, B>
+where
+    A: tower::Service<R>,
+    B: tower::Service<R, Response = A::Response, Error = A::Error>,
```
Comment (olix0r) [inline linkerd/stack/src/switch_ready.rs:94]: note that if you use `tower::util::Either`, you don't need to require that the error types are the same (though they both need to be `Into<Error>`)

**Review by olix0r — APPROVED** [review]
hooray, this looks great

File: linkerd/stack/src/switch_ready.rs (hunk 2 above)
Comment (hawkw) [reply linkerd/stack/src/switch_ready.rs:10]: whoopsy!

File: linkerd/stack/src/switch_ready.rs (hunk 3 above)
Comment (hawkw) [reply linkerd/stack/src/switch_ready.rs:94]: didn't realize `Either` was `Into<Error>` when both arms are `Into<Error>`...mainly i thought i'd avoid adding a `map_err` here that could be added in a separate layer?

Comment (hawkw) [reply linkerd/stack/src/switch_ready.rs:94]: oh, i see, it has a future impl, that's neat

**Pull request merged**